src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskSerializer.scala

```diff
@@ -77,50 +90,154 @@ object TaskSerializer {
       }
     }
 
-    Task(
+    constructTask(
       taskId = Task.Id(proto.getId),
       agentInfo = agentInfo,
-      reservationWithVolumes = reservationWithVolume,
-      launched = launchedTask
+      reservation,
+      reservedState,
+      launchedTask
     )
   }
 
-  def toProto(taskState: Task): Protos.MarathonTask = {
-    val builder = Protos.MarathonTask.newBuilder()
+  private[this] def constructTask(
+    taskId: Task.Id,
+    agentInfo: Task.AgentInfo,
+    reservationOpt: Option[ReservationWithVolumes],
+    reservedStateOpt: Option[Task.Reserved.State],
+    launchedOpt: Option[Task.Launched]): Task = {
 
-    builder.setId(taskState.taskId.idString)
+    (taskId, agentInfo, reservationOpt, reservedStateOpt, launchedOpt) match {
 
-    import taskState.agentInfo
-    builder.setHost(agentInfo.host)
-    agentInfo.agentId.foreach { agentId =>
-      builder.setSlaveId(MesosProtos.SlaveID.newBuilder().setValue(agentId))
-    }
-    builder.addAllAttributes(agentInfo.attributes.asJava)
+      case (_, _, Some(reservation), None, Some(launched)) =>
```

Comment:
why match against `taskId` and `agentInfo` if you don't use it then in the match?